blog/2024-11-22-inspektor-gadget-unittesting-framework.md

```diff
@@ -0,0 +1,255 @@
+}
+```
+
+## Defining Event Generating Function's Execution
+
+The unit-testing framework provide 2 option to for running the function 
+- `beforeGadgetRun`: This function is executed before the gadget starts running, It can be use to execute sometime before the gadget is started. also to test something snapshot gadgets.
+- `onGadgetRun`: This function is executed while the gadget is running When to Use:
+When gadget is used to trace a generated event while the gadget is running.
+
+## Configuring Gadget Options
+Before running the gadget, we define its configuration using GadgetRunnerOpts. This includes the gadget’s image, timeout, and optional namespace filters.
+
+```go
+opts := gadgetrunner.GadgetRunnerOpts[ExpectedTraceOpenEvent]{
+    Image:          "trace_open",
+    Timeout:        5 * time.Second,
+    MntnsFilterMap: mntnsFilterMap,
+    OnGadgetRun:    onGadgetRun,
+}
+gadgetRunner := gadgetrunner.NewGadgetRunner(t, opts)
+```
+#### Key Points:
+- Image: Specifies the gadget to be tested (e.g., "trace_open").
+- Timeout: Limits execution time to avoid infinite loops or hangs.
+- NormalizeEvents: The normalize function sets random fields to zero. Users handle gadget-specific fields, Various functions provided by Inspektor Gadget can be used for different tasks.
+- Namespace Filtering: Filters execution based on test-specific namespace configurations.
+- Paramvalues: this can be used to define gadget specific values such as `map-fetch-interval`, example can be found with in `top_file` unit test.
+
+### Running Gadget with given Gadget Options.
+after defining all the gadget options that are available a new gadgetRunner instance can be created and used to run gadget.
+```go
+gadgetRunner := gadgetrunner.NewGadgetRunner(t, opts)
+gadgetRunner.RunGadget()
+```
+
+### Validating captured output against the expected output 
+
+For every particular testCase a specific validation function can be defined in similar fashion:
+
+```go
+validateEvent: func(t *testing.T, info *utilstest.RunnerInfo, fd int, events []ExpectedTraceOpenEvent) {
+        utilstest.ExpectOneEvent(func(info *utilstest.RunnerInfo, fd int) *ExpectedTraceOpenEvent {
+            return &ExpectedTraceOpenEvent{
+                Proc:  info.Proc,
+                Fd:    uint32(fd),
+                FName: "/dev/null",
+            }
+        })(t, info, fd, events)
+}
+```
+
+In the given example, the test expects only one event to be captured. Similarly, other functions like `ExpectNoEvent` and `ExpectAtLeastOneEvent` can be used. For checking the equality of two fields, `require.Equal` can be utilized along with other supported functions from the require package.
+
+## Running the Test
+On normal testing machine we can use straight forward commands.
+```go
+# [Optional] If running the test for a gadget whose image resides in a remote container registry
+export GADGET_REPOSITORY=ghcr.io/my-org GADGET_TAG=latest
+
+# Run the test
+go test -exec 'sudo -E' -v
+```
```

Comment:
The `Selecting the Testing Environment:` section above can be placed around here.